### PR TITLE
Added support for ICT Modular Power System and ICT Sine Wave Inverter

### PR DESCRIPTION
**Pull request by noaheroufus**
DO NOT DELETE THIS TEXT

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)
- [ ] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

**Comment by noaheroufus** [comment]
I've visited the cla-assistant.io link and signed the CLA, but it doesn't seem to be carrying over to GitHub that I've signed it. When I visit the link again, it says I've already signed it.

**Comment by Jellyfrog** [comment]
> I've visited the cla-assistant.io link and signed the CLA, but it doesn't seem to be carrying over to GitHub that I've signed it. When I visit the link again, it says I've already signed it.

It's because you didn't commit with an email registered to your github account

File: includes/discovery/sensors/current/ict-mps.inc.php [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+// Output Current
+$outputCurrent = trim(snmp_get($device, 'outputCurrent.0', '-Oqv', 'ICTICT-MODULAR-POWER-SYSTEM-MIB'), '" ');
+if (!empty($outputCurrent)) {
+    $divisor = 1;
+    $index = 0;
+    $oid = '.1.3.6.1.4.1.39145.13.8.0';
+    $descr = 'Output Current (A)';
+    $type = 'ict-mps';
+    $currentValue = $outputCurrent / $divisor;
```
Comment (laf) [inline includes/discovery/sensors/current/ict-mps.inc.php]: You shouldn't need to apply the divisor before hand, that should be taken care of globally.

Any reason this isn't done as yaml?

File: includes/discovery/sensors/load/ict-swi.inc.php [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+// Inverter Load
+$inverterLoad = (int)(snmp_get($device, 'inverterPower.0', '-Oqv', 'ICT-SINE-WAVE-INVERTER-MIB'));
+if (!empty($inverterLoad)) {
+    $divisor = 1;
+    $index = 0;
+    $oid = '.1.3.6.1.4.1.39145.12.8.0';
+    $descr = 'Inverter Load';
+    $type = 'ict-swi';
+    $currentValue = $inverterLoad / $divisor;
```
Comment (laf) [inline includes/discovery/sensors/load/ict-swi.inc.php]: Same for this one.

File: includes/polling/os/ict-mps.inc.php [hunk 3]
```diff
@@ -0,0 +1,7 @@
+<?php
+
+//SNMPv2-SMI::enterprises.39145.13.1.0 = STRING: "Modular Power Series 48V"
+//SNMPv2-SMI::enterprises.39145.13.4.0 = STRING: "1.04"
+
+$hardware = trim(snmp_get($device, '1.3.6.1.4.1.39145.13.1.0', '-OQv', '', ''), '" ');
```
Comment (laf) [inline includes/polling/os/ict-mps.inc.php]: As you have MIBs for these devices can you not use named OIDs here? Also needs to be snmp_multi_get.

File: includes/polling/os/ict-swi.inc.php [hunk 4]
```diff
@@ -0,0 +1,7 @@
+<?php
+
+//SNMPv2-SMI::enterprises.39145.12.1.0 = STRING: "ICT1500-48SW"
+//SNMPv2-SMI::enterprises.39145.12.4.0 = STRING: "1.03"
+
+$hardware = trim(snmp_get($device, '1.3.6.1.4.1.39145.12.1.0', '-OQv', '', ''), '" ');
```
Comment (laf) [inline includes/polling/os/ict-swi.inc.php]: Same here.

File: includes/discovery/sensors/current/ict-mps.inc.php (hunk 1 above)
Comment (noaheroufus) [reply includes/discovery/sensors/current/ict-mps.inc.php]: Removed unnecessary math. No particular reason it's not done in yaml.

File: includes/discovery/sensors/load/ict-swi.inc.php (hunk 2 above)
Comment (noaheroufus) [reply includes/discovery/sensors/load/ict-swi.inc.php]: Removed unnecessary math.

File: includes/polling/os/ict-mps.inc.php (hunk 3 above)
Comment (noaheroufus) [reply includes/polling/os/ict-mps.inc.php]: Moved to snmp_get_multi.

File: includes/polling/os/ict-swi.inc.php (hunk 4 above)
Comment (noaheroufus) [reply includes/polling/os/ict-swi.inc.php]: Moved to snmp_get_multi.

**Comment by noaheroufus** [comment]
Anything else I'm missing or need tweaking before this is ready to merge?

**Comment by Jellyfrog** [comment]
Still missing tests it seems like, take a look at https://docs.librenms.org/Developing/os/Test-Units

**Comment by noaheroufus** [comment]
Believe those should be the required tests? Please let me know if otherwise.

File: includes/discovery/sensors/state/ict-swi.inc.php [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        "state_name" => "transferRelayStatus",
+        "states" => [
+            ['value' => 1, 'generic' => 1, 'graph' => 0, 'descr' => 'inverter'],
+            ['value' => 2, 'generic' => 0, 'graph' => 0, 'descr' => 'grid'],
+        ]
+    ]
+);
+
+foreach ($oids as $oid => $info) {
+    $data = (int)(snmpget($device, $oid, '-0qv', 'ICT-SINE-WAVE-INVERTER-MIB'));
```
Comment (Jellyfrog) [inline includes/discovery/sensors/state/ict-swi.inc.php]:
```suggestion
    $data = (int)(snmp_get($device, $oid, '-0qv', 'ICT-SINE-WAVE-INVERTER-MIB'));
```

**Comment by Jellyfrog** [comment]
Most of this could be written in the yaml format instead

**Comment by noaheroufus** [comment]
OK, so are you saying drop this PR and recreate with yaml instead? I didn't have much luck with the graphs populating in my last attempt, but perhaps I missed something.

**Comment by Jellyfrog** [comment]
Well if you want to give it a shot it would be nice. You can commit in the same branch. No need to close it and reopen

**Comment by noaheroufus** [comment]
Alright, I found my way around the yaml sensors. Here's a commit with the legacy code removed in place of yaml.

**Comment by PipoCanaja** [comment]
Nicely done. Ready for merge. Thanx !

**Comment by murrant** [comment]
This pull request has been mentioned on **LibreNMS Community**. There might be relevant details there:

https://community.librenms.org/t/v1-63-release-changelog-april-2020/11828/1

**Pull request merged**